core/lib/libmbedtls/src/tee_lmd_provider.c

```diff
@@ -583,53 +683,290 @@ crypto_acipher_ecc_shared_secret(struct ecc_keypair *private_key __unused,
  ******************************************************************************/
 
 #if defined(_CFG_CRYPTO_WITH_CIPHER)
-TEE_Result crypto_cipher_alloc_ctx(void **ctx __unused, uint32_t algo __unused)
+
+static TEE_Result cipher_get_ctx_size(uint32_t algo, size_t *size)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	switch (algo) {
+#if defined(CFG_CRYPTO_ECB)
+	case TEE_ALG_AES_ECB_NOPAD:
+		*size = sizeof(mbedtls_cipher_context_t);
+		break;
+#if defined(CFG_CRYPTO_CBC)
+#endif
+	case TEE_ALG_AES_CBC_NOPAD:
+		*size = sizeof(mbedtls_cipher_context_t);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CTR)
+	case TEE_ALG_AES_CTR:
+		*size = sizeof(mbedtls_cipher_context_t);
+		break;
+#endif
+#if defined(CFG_CRYPTO_XTS)
+	case TEE_ALG_AES_XTS:
+		return TEE_ERROR_NOT_SUPPORTED;
+#endif
+#if defined(CFG_CRYPTO_CTS)
+	case TEE_ALG_AES_CTS:
+		return TEE_ERROR_NOT_SUPPORTED;
+#endif
+#if defined(CFG_CRYPTO_ECB)
+	case TEE_ALG_DES_ECB_NOPAD:
+	case TEE_ALG_DES3_ECB_NOPAD:
+		*size = sizeof(mbedtls_cipher_context_t);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC)
+	case TEE_ALG_DES_CBC_NOPAD:
+	case TEE_ALG_DES3_CBC_NOPAD:
+		*size = sizeof(mbedtls_cipher_context_t);
+		break;
+#endif
+	default:
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	return TEE_SUCCESS;
 }
 
-void crypto_cipher_free_ctx(void *ctx, uint32_t algo __unused)
+TEE_Result crypto_cipher_alloc_ctx(void **ctx_ret, uint32_t algo)
 {
-	if (ctx)
-		assert(0);
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+	int lmd_res;
+	TEE_Result res;
+	size_t ctx_size;
+	void *ctx;
+
+	switch (algo) {
```

Comment:
This `switch()` could be structured a bit more efficiently, like what I'm asking for in `cipher_get_ctx_size()` above.